src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectWorkspaceStateGenerator.cs

```diff
@@ -109,28 +62,60 @@ public void Dispose()
         _semaphore.Release();
         _semaphore.Dispose();
 
-        BlockBackgroundWorkStart?.Set();
+        _blockBackgroundWorkStart?.Set();
     }
 
-    private async Task UpdateWorkspaceStateAsync(Project? workspaceProject, IProjectSnapshot projectSnapshot, CancellationToken cancellationToken)
+    public void EnqueueUpdate(Project? workspaceProject, IProjectSnapshot projectSnapshot)
     {
-        // We fire this up on a background thread so we could have been disposed already, and if so, waiting on our semaphore
-        // throws an exception.
-        if (_disposed)
+        if (_disposeTokenSource.IsCancellationRequested)
         {
             return;
         }
 
-        // Specifically not using BeginBlock because we want to capture cases where tag helper discovery never finishes.
-        var telemetryId = Guid.NewGuid();
-        _telemetryReporter.ReportEvent("taghelperresolve/begin", Severity.Normal,
-            new Property("id", telemetryId),
-            new Property("tagHelperCount", projectSnapshot.ProjectWorkspaceState.TagHelpers.Length));
+        lock (_updates)
+        {
+            if (_updates.TryGetValue(projectSnapshot.Key, out var updateItem))
+            {
+                if (!updateItem.UpdateTask.IsCompleted &&
+                    !updateItem.IsCancellationRequested)
+                {
+                    _logger.LogTrace($"Cancelling previously enqueued update for '{projectSnapshot.FilePath}'.");
+                }
+
+                updateItem.Dispose();
+            }
+
+            _logger.LogTrace($"Enqueuing update for '{projectSnapshot.FilePath}'");
+
+            _updates[projectSnapshot.Key] = new UpdateItem(
```

Comment:
Nit: It feels a little hidden that this actually starts the work. Everything talks about "enqueing" but there is no queue processing that is visible. I think a comment above this, or perhaps using a static `UpdateItem.CreateAndStartWork` method rather an a constructor would make it clearer.

Reply:
Great point -- I'll do that.